pyro/distributions/sine_skewed.py

```diff
@@ -0,0 +1,92 @@
+from math import pi
+
+import torch
+from torch.distributions import Uniform
+
+from pyro.distributions import constraints
+
+from .torch_distribution import TorchDistribution
+
+
+class SineSkewed(TorchDistribution):
+    """The Sine Skewed distribution [1] is a distribution for breaking pointwise-symmetry on a base-distribution over
+    the d-dimensional torus.
+
+    This distribution requires the base distribution on a torus. The parameter skewness can be inferred using
+    :class:`~pyro.infer.HMC` or :class:`~pyro.infer.NUTS`. For example, the following will produce a uniform prior
+    over skewness for the 1-torus,::
+
+        def model(...):
+            ...
+            skewness_phi = pyro.sample(f'skewness_phi', Uniform(skewness.abs().sum(), 1 - tots))
+            psi_bound = 1 - skewness_phi.abs()
+            skewness_psi = pyro.sample(f'skewness_psi', Uniform(-psi_bound, psi_bound)
+            skewness = torch.stack((skewness_phi, skewness_psi), dim=0)
+            ...
+
+    .. note:: An event in the base distribution must be on a d-torus, so the event_shape must be (d, 2) or (2,).
+
+    .. note:: For the skewness parameter, it must hold that the sum of the absolute value of its weights for an event
+        must be less than or equal to one. See eq. 2.1 in [1].
+
+    ** References: **
+      1. Sine-skewed toroidal distributions and their application in protein bioinformatics
+         Ameijeiras-Alonso, J., Ley, C. (2019)
+
+    :param base_density: base density on a d-dimensional torus.
+    :param skewness: skewness of the distribution.
+    """
+    arg_constraints = {'skewness': constraints.independent(constraints.interval(-1., 1.), 1)}
```

Comment:
Unless you split this distribution into `SineSkewedCircular` and `SineSkewedTorus` for arbitrary event_dim, the constraint here is `constraints.dependent`. You can add some assertation to the constructor.

Reply:
I'll split it into two distributions; I need to check that the distribution is proper when over an n-sphere. As you've already pointed out, this will make the implementation much more straightforward.